Memory Manager logging improvements

swatisehgal opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

#### What type of PR is this?
/kind cleanup
<!--
Add one of the following kinds:
/kind bug
/kind cleanup
/kind documentation
/kind feature

Optionally add one or more of the following kinds if applicable:
/kind api-change
/kind deprecation
/kind failing-test
/kind flake
/kind regression
-->

#### What this PR does / why we need it:

Enhancing Memory manager logs to ensure consistency across resource managers. This PR adds logs, some missing metadata to logs, updates log levels to prevent excessive noise and makes use of klog.InfoS() and klog.ErrorS() as intended.

#### Which issue(s) this PR fixes:

<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Partially Fixes #123037

#### Special notes for your reviewer:

We are doing memory manager logging cleanup first. This will be followed by migration of memory manager to contextual logging.  

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs

```

swatisehgal committed:
node: memory-mgr: Change ErrorS(nil, ...) to InfoS

Ensure consistency across resource managers and update
ErrorS(nil, ...) to InfoS.

Signed-off-by: Swati Sehgal <swsehgal@redhat.com>

swatisehgal committed:
node: memory-mgr: Add metadata to logs

Ensure that whereever possible, we provide sufficient metadata
inclusing pod name and UID to allow filtering by pod name or its
UID.

Signed-off-by: Swati Sehgal <swsehgal@redhat.com>

swatisehgal committed:
node: memory-mgr: Update log levels and add logs to capture state update

Signed-off-by: Swati Sehgal <swsehgal@redhat.com>

swatisehgal committed:
node: memory-mgr: Update klog.Infof(..., err) to klog.ErrorS(err,...)

Signed-off-by: Swati Sehgal <swsehgal@redhat.com>

swatisehgal commented:
/cc @ffromani

swatisehgal committed:
node:memory-mgr: Add logs on the happy path

We have reasonable amount of logs when things go wrong.
While debugging, it can be useful to have logs to indicate that
things have gone as expected especially when it comes to
important events like successful startup of memory manager
and successful allocation of resources.

Signed-off-by: Swati Sehgal <swsehgal@redhat.com>

swatisehgal committed:
node: memory-mgr: Add logs when memory allocation is skipped due to QoS

Signed-off-by: Swati Sehgal <swsehgal@redhat.com>

ffromani commented on pkg/kubelet/cm/memorymanager/memory_manager.go:
```diff
@@ -248,11 +248,11 @@ func (m *manager) GetMemoryNUMANodes(pod *v1.Pod, container *v1.Container) sets.
 
 	if numaNodes.Len() == 0 {
-		klog.V(5).InfoS("No allocation is available", "pod", klog.KObj(pod), "containerName", container.Name)
+		klog.V(5).InfoS("No allocation is available", "pod", klog.KObj(pod), "podUID", pod.UID, "containerName", container.Name)
```
While I'm in favor of this change, I can't help wondering if logging _both_ pod namespace/name and uid is not too much.
I think we need both to correctly correlate, but adding always both feels heavy-handed.
When touching logs, adding is almost always easier than removing, so in general I'd be prudent to add stuff, even stuff which is likely to be useful.

I'm still making  my mind as I face similar dilemmas in my https://github.com/kubernetes/kubernetes/pull/125912

swatisehgal replied:
Sure, I understand. My thought process was that one of the key features of structural logging is to allow users to filter and because logs are sprinkled with both (albeit not always in the same log) the user might choose to use one over the other and we should allow that. Having said that, I am open to suggestions in this area. One thing I want to ensure is consistency across the managers so let's perhaps agree to _a_ way and stick to that in all the managers.

ffromani commented on pkg/kubelet/cm/memorymanager/state/state_mem.go:
```diff
@@ -94,6 +94,7 @@ func (s *stateMemory) SetMemoryAssignments(assignments ContainerMemoryAssignment
 	defer s.Unlock()
 
 	s.assignments = assignments.Clone()
+	klog.InfoS("Updated Memory assignments", "assignments", assignments)
```
assigments can be large, because of this I'd do this with `V(5)` at least

swatisehgal replied:
Ack, will update.

ffromani commented on pkg/kubelet/cm/memorymanager/memory_manager.go:
```diff
@@ -205,6 +205,7 @@ func (m *manager) Start(activePods ActivePodsFunc, sourcesReady config.SourcesRe
 
 	m.allocatableMemory = m.policy.GetAllocatableMemory(m.state)
 
+	klog.V(4).InfoS("memorymanager started successfully", "policy", m.policy.Name())
```
similar comments to https://github.com/kubernetes/kubernetes/pull/129685/files/110868691b271a7448e613f10032b150d23f429e..080962a3a788c54d7f272947d5f4f61903875520#diff-c11864857fa45d5c2d29c3a85d753ccd2db31cca1c7fa046b7144e71f6e59029

ffromani commented on pkg/kubelet/cm/memorymanager/memory_manager.go:
```diff
@@ -269,6 +270,8 @@ func (m *manager) Allocate(pod *v1.Pod, container *v1.Container) error {
 		return err
 	}
+
+	klog.V(4).InfoS("Memory Allocated by memorymanager successfully", "pod", klog.KObj(pod), "podUID", pod.UID, "containerName", container.Name)
```
ditto

swatisehgal replied:
Moved to static policy but just not sure if it makes sense to add `containerBlocks` to the logs to capture the allocated memory. I am leaning towards not...

ffromani commented on pkg/kubelet/cm/memorymanager/policy_static.go:
```diff
@@ -97,6 +97,7 @@ func (p *staticPolicy) Start(s state.State) error {
 func (p *staticPolicy) Allocate(s state.State, pod *v1.Pod, container *v1.Container) (rerr error) {
 	// allocate the memory only for guaranteed pods
 	if v1qos.GetPodQOS(pod) != v1.PodQOSGuaranteed {
+		klog.V(5).InfoS("Memory Allocation Skipped for pod as it is allocated only for Guaranteed pods", "pod", klog.KObj(pod), "podUID", pod.UID, "containerName", container.Name)
```
similar comments to https://github.com/kubernetes/kubernetes/pull/129685/commits/9d46064fa7f2321b570ae8d2bc8e28be264af0db#r1920259412

ffromani reviewed (commented):
thanks! overall LGTM, with some points shared with https://github.com/kubernetes/kubernetes/pull/129685/ - let's maybe sort out in one place, the other can be trivially adapted

swatisehgal commented:
/test pull-kubernetes-e2e-kind

bart0sh commented:
/triage accepted

Tal-or commented on pkg/kubelet/cm/memorymanager/policy_static.go:
```diff
@@ -304,24 +304,24 @@ func regenerateHints(pod *v1.Pod, ctn *v1.Container, ctnBlocks []state.Block, re
 
 		if b.Size != reqRsrc[b.Type] {
-			klog.InfoS("Memory already allocated with different numbers than requested", "podUID", pod.UID, "type", b.Type, "containerName", ctn.Name, "requestedResource", reqRsrc[b.Type], "allocatedSize", b.Size)
+			klog.InfoS("Memory already allocated with different numbers than requested", "pod", klog.KObj(pod), "podUID", pod.UID, "type", b.Type, "containerName", ctn.Name, "requestedResource", reqRsrc[b.Type], "allocatedSize", b.Size)
```
NIT: it worth keep logging the variables in consistent order so `containerName` comes before `type`

swatisehgal replied:
Yeah, good point. Will update. Thanks!

Tal-or commented on pkg/kubelet/cm/memorymanager/policy_static.go:
```diff
@@ -304,24 +304,24 @@ func regenerateHints(pod *v1.Pod, ctn *v1.Container, ctnBlocks []state.Block, re
 		containerNUMAAffinity, err := bitmask.NewBitMask(b.NUMAAffinity...)
 		if err != nil {
-			klog.ErrorS(err, "Failed to generate NUMA bitmask")
+			klog.ErrorS(err, "Failed to generate NUMA bitmask", "pod", klog.KObj(pod), "podUID", pod.UID, "type", b.Type, "containerName", ctn.Name)
```
ditto

Tal-or reviewed (commented):
Really nice, make logs much clearer now. small comments inside

swatisehgal committed:
node: memory-mgr: Adhere to the message style guidelines

Ensure that the log messages adhere to the message style guildelines
as captured [here](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-instrumentation/logging.md#message-style-guidelines).

Signed-off-by: Swati Sehgal <swsehgal@redhat.com>

ffromani commented:
/priority important-longterm

prerequisite for the contextual logging work, which is in turn important-longterm
LGTM

ffromani commented:
/approve
